graphix/parameter.py

```diff
@@ -0,0 +1,921 @@
+        return Log(self).simplify()
+
+    def conjugate(self) -> Expression | complex:
+        if self.is_float():
+            return self
+        return Conjugate(self)
+
+    def sqrt(self) -> Expression | complex:
+        return Sqrt(self).simplify()
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __str__(self) -> str:
+        target = []
+        self.format(0, target)
+        return "".join(target)
+
+    def format(self, precedence: int, target: list[str]) -> None: ...
+
+    def subs(self, variable: Parameter, value: Expression | numbers.Number) -> Expression | complex: ...
+
+    def simplify(self) -> Expression | complex: ...
+
+    def is_float(self) -> bool:
+        return False
+
+    def is_integer(self) -> bool:
+        return False
+
+    def flatten(self) -> Expression:
+        return self
+
+    def conj(self) -> Expression | complex:
+        return self.conjugate()
+
+
+class Parameter(Expression):
+    """Placeholder for measurement angles, which allows the pattern optimizations
+    without specifying measurement angles for measurement commands.
+    Either use for rotation gates of :class:`Circuit` class or for
+    the measurement angle of the measurement commands to be added with :meth:`Pattern.add` method.
+    Example:
+    .. code-block:: python
+
+        import numpy as np
+        from graphix import Circuit
+        circuit = Circuit(1)
+        alpha = Parameter('alpha')
+        # rotation gate
+        circuit.rx(0, alpha)
+        pattern = circuit.transpile()
+        # Both simulations (numeric and symbolic) will use the same
+        # seed for random number generation, to ensure that the
+        # explored branch is the same for the two simulations.
+        seed = np.random.integers(2**63)
+        # simulate with parameter assignment
+        sv = pattern.subs(alpha, 0.5).simulate_pattern(pr_calc=False, rng=np.random.default_rng(seed))
+        # simulate without pattern assignment
+        # (the resulting state vector is symbolic)
+        # Note: pr_calc=False is mandatory since we cannot compute probabilities on
+        # symbolic states; we explore one arbitrary branch.
+        sv2 = pattern.simulate_pattern(pr_calc=False, rng=np.random.default_rng(seed))
+        # Substituting alpha in the resulting state vector should yield the same result
+        assert np.allclose(sv.psi, sv2.subs(alpha, 0.5).psi)
+    """
+
+    def __init__(self, name: str, type: type = float) -> None:
+        """Create a new :class:`Parameter` object.
+
+        Parameters
+        ----------
+        name : str
+            name of the parameter, used for binding values.
+        """
+        self.__name = name
+        self.__type = type
+
+    @property
+    def name(self) -> str:
+        return self.__name
+
+    @property
+    def type(self) -> type:
+        return self.__type
+
+    def subs(self, variable: Parameter, value: Expression | numbers.Number) -> Expression | complex:
```

Comment:
I once here that `numbers.XXX` does not work efficiently with modern typing systems (see Fluent Python 2nd ed.).

Do you come up with alternatives to that?